dashboard/app/models/concerns/partial_registration.rb

```diff
@@ -8,16 +8,64 @@ module PartialRegistration
   extend ActiveSupport::Concern
 
   USER_ATTRIBUTES_SESSION_KEY = 'devise.user_attributes'
-  OAUTH_PARAMS_TO_STRIP = %w(oauth_token oauth_refresh_token)
+
+  # A common error during registration, especially when OAuth is involved, is a
+  # Cookie Overflow.  This module stores attributes for a partially-created user
+  # in the _learn_session cookie, which cannot exceed 4KB.  In fact, it we seem
+  # to encounter issues when the cookie is more than 2KB, since Devise and other
+  # parts of our application may also be using the session.
+  # To work around this, we keep known large values out of the session and
+  # instead store them in the shared application cache.
+  ATTRIBUTES_TO_KEEP_IN_CACHE = %w(
+    school
+    full_address
+  )
+  # also these large values from the user serialized properties blob
+  PROPERTIES_TO_KEEP_IN_CACHE = %w(
+    oauth_token
+    oauth_refresh_token
+  )
+  # We also drop some user model attributes that we know we don't need to preserve
+  # if we haven't created a user yet.
+  ATTRIBUTES_TO_DROP = %w(
+    id
+    studio_person_id
+    reset_password_token
+    reset_password_sent_at
+    remember_created_at
+    sign_in_count
+    current_sign_in_at
+    last_sign_in_at
+    current_sign_in_ip
+    last_sign_in_ip
+    created_at
+    updated_at
+    username
+    admin
+    total_lines
+    secret_picture_id
+    secret_words
+    active
+    deleted_at
+    purged_at
+    invitation_token
+    invitation_created_at
+    invitation_sent_at
+    invitation_accepted_at
+    invitation_limit
+    invited_by_id
+    invited_by_type
+    invitations_count
+  )
```

Comment:
> do we have a sense of, before and after this change, how big the user attributes are that this feature is storing in the session cookie?

I have estimates but not measurements yet.

**Before (estimated):** By taking example `devise.user_attributes` from the session as reported in example Honeybadger errors, swapping in realistic values for e.g. the hashed values that HB hides, and removing any formatting, I can determine that the typical unencrypted size of keys and values in this part of the session is about 1.1KB, about a quarter of the available space.

**After (estimated):** By taking the same example and removing key-value pairs that would be omitted by this code change, the unencrypted size of keys and values drops to about 0.5KB.

It's certainly not a deep fix, but an extra 0.6KB of clearance might make a big difference vs a 4KB limit.

I haven't done a deep dive yet into exactly what transformations happen to this data on its way to the session cookie, affecting its final effective size.  I will follow up on this.

Reply:
Thanks for these details. Based on this, one alternative which comes to mind is just to use a different cookie besides _learn_session for this codepath. 

Depending on the persistency requirements of this data, this might be more or less complexity than the current solution, for example if you need to wipe this cookie on logout then it might add some complexity.

On the plus side, a separate cookie would save an additional 600B (out of 4000B), and isolate any session data usage problems between this codepath and others, both of which seem like big wins, especially in an area where we don't really have much insight into exactly what's going wrong.

Reply:
Separate cookies seems like a reasonable solution for both PartialRegistration and callouts_seen.  Some open questions we need to address as part of that change:

- PartialRegistration may contain sensitive information, and it's probably a good idea to encrypt it, much like how the session cookie is currently encrypted.  What is the best way to do this?
- Some older browsers [enforce a 4KB total size limit for _all_ cookies on a domain](http://browsercookielimits.squawky.net/). It looks like this is unlikely to be an issue for our support matrix (IE11 has a limit of about 12KB, other modern browsers have effectively eliminated this limit) but I'd like to be sure.

Reply:
I'm not sure if we would ever set both the rails session AND the callouts_seen or PartialRegistration cookies at the same time. but if we did, this thread seems relevant: https://stackoverflow.com/questions/52203972/maximum-cookie-size-of-current-browsers-year-2018#comment96057193_52492934

> I get warnings in my Chrome developer console when the total size of all Set-Cookie headers in a response exceeds 4096 characters.

Reply:
Ooh, great catch.  I don't think we ever set all three of these things (PartialRegistration, callouts_seen, the rest of the session) in the same response right now, but for the purposes of this design I'll be pessimistic and assume some future change could introduce this behavior without our noticing.